### ISSUE TITLE
FCOS version of DDOD

### ISSUE THREAD

**Issue by memoiry** [zehuichen123/DDOD#3]
Hi, thanks for your great work!.

It seems that there is no FCOS version of DDOD (Use anchor point instead of anchor) implemented as indicated in the Table 3 of your paper, could you also share that part of code and config for reference?

Thanks!

**Comment by zehuichen123** [comment]
Actually, the difference between ATSS and the advanced version of FCOS is the regression type (regress from point or regress from anchor). Therefore, the FCOS in our code is to simply replace the original bbox_coder from:
```
# bbox_coder=dict(
        #     type='DeltaXYWHBBoxCoder',
        #     target_means=[.0, .0, .0, .0],
        #     target_stds=[0.1, 0.1, 0.2, 0.2]),
```
to 
```
 bbox_coder=dict(
            type='TBLRCenterCoder',
            normalizer=1/8.,
            normalize_by_wh=True
        ), 
```

**Comment by memoiry** [comment]
Interesting, do you mean replace bbox_coder in `coco_cfg/ddod_r50_1x.py` could reproduce advanced version of FCOS?

Also I didn't find `TBLRCenterCoder ` in the code, I just see `TBLRBBoxCoder`. Any suggestion?

Thanks!

**Comment by memoiry** [comment]
I just found your latest commit, please ignore my comment.

**Comment by memoiry** [comment]
Hi, @zehuichen123 I just checked the `TBLRCenterCoder` implementation and found that it uses `normalize_by_wh` as a normalizer and there is anchor size in the anchors, is that normal? I mean, anchor point should not have anchor size?

**Comment by memoiry** [comment]
Also, I found that anchor point generated from anchor_generator is different from FCOS.

if we have a  2x2 feature map of stride 8. For anchor generator.

```python
In [20]: self = AnchorGenerator(ratios=[1.0],
    ...:             octave_base_scale=8,
    ...:             scales_per_octave=1,
    ...:             strides=[8])

In [21]: all_anchors = self.grid_anchors([(2, 2)], device='cpu')

In [34]: (all_anchors[0][:, 0:2] + all_anchors[0][:, 2:4]) / 2
Out[34]:
tensor([[0., 0.],
        [8., 0.],
        [0., 8.],
        [8., 8.]])
```
4 anchor point centered at 

tensor([[0., 0.],
        [8., 0.],
        [0., 8.],
        [8., 8.]])

However, for FCOS

```python
In [24]:     def compute_locations_per_level(h, w, stride, device):
    ...:         shifts_x = torch.arange(
    ...:             0, w * stride, step=stride,
    ...:             dtype=torch.float32, device=device
    ...:         )
    ...:         shifts_y = torch.arange(
    ...:             0, h * stride, step=stride,
    ...:             dtype=torch.float32, device=device
    ...:         )
    ...:         shift_y, shift_x = torch.meshgrid(shifts_y, shifts_x)
    ...:         shift_x = shift_x.reshape(-1)
    ...:         shift_y = shift_y.reshape(-1)
    ...:         locations = torch.stack((shift_x, shift_y), dim=1) + stride // 2
    ...:         return locations
    ...:

In [25]: compute_locations_per_level(2,2,8,'cpu')
Out[25]:
tensor([[ 4.,  4.],
        [12.,  4.],
        [ 4., 12.],
        [12., 12.]])
```
Which is quite different.

Do you have any suggestion? Thanks!

**Comment by zehuichen123** [comment]
Yes, this normalization is actually a constant for each FPN level (8 for P3, 16 for P4...), which can be integrated into the learnable variable `scale`. But we empirically found normalized with w/h yields better results in widerface experiments.
As for the anchor generation process, this difference does not matter much, maybe you can have a try :). I think the main differences between ATSS and FCOS lie in two parts: label assignment strategy and regression type(regress from anchor / point), which is mentioned in ATSS paper.

**Comment by memoiry** [comment]
@zehuichen123 , Thanks for your quick reply! 

Another question is that do you have AP results on WiderFace for DDOD-FCOS and how it performs compared to DDOD-ATSS?

**Comment by zehuichen123** [comment]
According to our records, the performance of DDOD-FCOS is comparable with DDOD-ATSS (92.9 vs 93.0).

**Comment by memoiry** [comment]
Got it, thanks!
